cpp/src/Slice/MetadataValidator.cpp

```diff
@@ -0,0 +1,299 @@
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+bool
+Slice::MetadataValidator::visitInterfaceDefStart(const InterfaceDefPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    return true;
+}
+
+bool
+Slice::MetadataValidator::visitExceptionStart(const ExceptionPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    return true;
+}
+
+bool
+Slice::MetadataValidator::visitStructStart(const StructPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    return true;
+}
+
+void
+Slice::MetadataValidator::visitOperation(const OperationPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+void
+Slice::MetadataValidator::visitParamDecl(const ParamDeclPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+void
+Slice::MetadataValidator::visitDataMember(const DataMemberPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+void
+Slice::MetadataValidator::visitSequence(const SequencePtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    p->setTypeMetadata(validate(p->typeMetadata(), p->type()));
+}
+
+void
+Slice::MetadataValidator::visitDictionary(const DictionaryPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    p->setKeyMetadata(validate(p->keyMetadata(), p->keyType()));
+    p->setValueMetadata(validate(p->valueMetadata(), p->valueType()));
+}
+
+void
+Slice::MetadataValidator::visitEnum(const EnumPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+}
+
+void
+Slice::MetadataValidator::visitConst(const ConstPtr& p)
+{
+    p->setMetadata(validate(p->getMetadata(), p));
+    p->setTypeMetadata(validate(p->typeMetadata(), p->type()));
+}
+
+MetadataList
+Slice::MetadataValidator::validate(MetadataList metadata, const SyntaxTreeBasePtr& p)
+{
+    // Iterate through the provided metadata and check each one for validity.
+    for (MetadataList::const_iterator i = metadata.begin(); i != metadata.end(); ++i)
+    {
+        const MetadataPtr& meta = *i++;
+        if (auto message = validateMetadata(meta, p))
+        {
+            // If a warning message was returned from the validation function, it means the metadata was invalid.
+            // We remove it from the list and emit a warning to the user about it.
+            metadata.remove(meta);
+            p->unit()->warning(meta->file(), meta->line(), InvalidMetadata, *message);
+        }
+    }
+
+    return metadata;
+}
+
+optional<string>
+Slice::MetadataValidator::validateMetadata(const MetadataPtr& metadata, const SyntaxTreeBasePtr& p)
+{
+    // We only want to check metadata that that doesn't have a language prefix (ie. parser metadata).
+    const string& directive = metadata->directive();
+    if (directive.find(':') != string::npos)
+    {
+        return nullopt;
+    }
+
+    // Check each of the language-agnostic metadata directives that the parser is aware of.
+    if (directive == "amd")
```

Comment:
This looks a lot like the logic we have for dispatching operations: we receive an operation name ("op") and we need to select the correct handler that unmarshals the parameters and more. It would make sense to use the same "dispatch" logic. See for example:
https://github.com/zeroc-ice/ice/blob/40171af6057781f3fd07817c02875cebd525df54/cpp/src/Ice/Object.cpp#L103

Reply:
The operation dispatch logic is definitely more performant than my code,
but I also think it's much more complicated to follow, and there's more places for things to go wrong (we need to keep indices, the list, and it's length in sync)...

So, personally I'd prefer to keep the simpler logic I have here, since the Slice compiler's performance is a secondary concern.